File: 1-D Dynamic Programming/5 Longest Palindromic Substring.py
from functools import cache

class Solution:
    @cache
    def longestPalindrome(self, s: str) -> str:
        seen = {} # string : bool
        res = ""
        if len(s) == 1:
            return s[0]
        if len(s) == 2:
            return s[0] if s[0] != s[1] else s

        def validPalindrome(l: int, r: int) -> bool:
            while l < r:
                if s[l] == s[r]:
                    l += 1
                    r -= 1
                else:
                    return False
            return True
        def longest(l: int, r: int, currMax: int):
            if l == r or (l, r) in seen or len(s[l:r+1]) < currMax:
                return
            seen[(l, r)] = validPalindrome(l, r)
            longest(l+1, r, currMax)
            for k, v in seen.items():
                if v and len(s[k[0]:k[1]+1]) > currMax:
                    currMax = len(s[k[0]:k[1]+1])
            longest(l, r-1, currMax)
        longest(0, len(s)-1, 1)
        
        
        for k, v in seen.items():
            if v and (k[1] - k[0] + 1) > len(res):
                res = s[k[0]:k[1]+1]
        print(len(seen))
        return s[0] if res == "" else res
    # def longestPalindrome(self, s: str) -> str:
    #     for i in range(len(s)):
            
    
sol = Solution()
# print(sol.longestPalindrome("boqylncwfahjzvawrojyhqiymirlkfzkhtvmbjnbfjxzewqqqcfnximdnrxtrbafkimcqvuprgrjetrecqkltforcudmbpofcxqdcirnaciggflvsialdjtjnbrayeguklcbdbkouodxbmhgtaonzqftkebopghypjzfyqutytbcfejhddcrinopynrprohpbllxvhitazsjeyymkqkwuzfenhphqfzlnhenldbigzmriikqkgzvszztmvylzhbfjoksyvfdkvshjzdleeylqwsapapduxrfbwskpnhvmagkolzlhakvfbvcewvdihqceecqhidvwecvbfvkahlzlokgamvhnpkswbfrxudpapaswqlyeeldzjhsvkdfvyskojfbhzlyvmtzzsvzgkqkiirmzgibdlnehnlzfqhphnefzuwkqkmyyejszatihvxllbphorprnyponircddhjefcbtytuqyfzjpyhgpobektfqznoatghmbxdouokbdbclkugeyarbnjtjdlaisvlfggicanricdqxcfopbmducroftlkqcertejrgrpuvqcmikfabrtxrndmixnfcqqqwezxjfbnjbmvthkzfklrimyiqhyjorwavzjhafwcnlyqob"))
print(sol.longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth"))

File: 1-D Dynamic Programming/5 Longest Palindromic Substring.go
package main

import "fmt"

type Coords struct {
	l int
	r int
}

func longestPalindrome(s string) string {
	seen := map[Coords]bool{}
	res := ""
	if len(s) == 1 {
		return string(s[0])
	}
	if len(s) == 2 {
		if s[0] == s[1] {
			return s
		}
		return string(s[0])
	}
	var validPalindrome func(int, int) bool
	var longest func(int, int, int)

	validPalindrome = func(l, r int) bool {
		for l < r {
			if s[l] == s[r] {
				l += 1
				r -= 1
			} else {
				return false
			}
		}
		return true
	}

	longest = func(l, r, currMax int) {
		coords := Coords{
			l: l,
			r: r,
		}
		if l == r || len(s[l:r+1]) < currMax {
			return
		}
		if _, ok := seen[coords]; ok {
			return
		}
		seen[coords] = validPalindrome(l, r)
		for k, v := range seen {
			if v && len(s[k.l:k.r+1]) > currMax {
				currMax = len(s[k.l : k.r+1])
			}
		}
		longest(l+1, r, currMax)
		longest(l, r-1, currMax)
	}
	longest(0, len(s)-1, 0)

	for coords, val := range seen {
		if val && (coords.r-coords.l) >= len(res) {
			res = s[coords.l : coords.r+1]
		}
	}
	fmt.Println("Length of seen:", len(seen))
	if res == "" {
		return string(s[0])
	}
	return res
}

func main() {
	fmt.Println(longestPalindrome("cbbd"))
	fmt.Println(longestPalindrome("civilwartestingwhetherthatnaptionoranynartionsoconceivedandsodedicatedcanlongendureWeareqmetonagreatbattlefiemldoftzhatwarWehavecometodedicpateaportionofthatfieldasafinalrestingplaceforthosewhoheregavetheirlivesthatthatnationmightliveItisaltogetherfangandproperthatweshoulddothisButinalargersensewecannotdedicatewecannotconsecratewecannothallowthisgroundThebravelmenlivinganddeadwhostruggledherehaveconsecrateditfaraboveourpoorponwertoaddordetractTgheworldadswfilllittlenotlenorlongrememberwhatwesayherebutitcanneverforgetwhattheydidhereItisforusthelivingrathertobededicatedheretotheulnfinishedworkwhichtheywhofoughtherehavethusfarsonoblyadvancedItisratherforustobeherededicatedtothegreattdafskremainingbeforeusthatfromthesehonoreddeadwetakeincreaseddevotiontothatcauseforwhichtheygavethelastpfullmeasureofdevotionthatweherehighlyresolvethatthesedeadshallnothavediedinvainthatthisnationunsderGodshallhaveanewbirthoffreedomandthatgovernmentofthepeoplebythepeopleforthepeopleshallnotperishfromtheearth"))
}
